server/drivers/hd44780-gpio.c

```diff
@@ -0,0 +1,352 @@
+/** \file server/drivers/hd44780-gpio.c
+ * \c GPIO connection type of \c hd44780 driver for Hitachi HD44780 based LCD
+ * displays.
+ *
+ * The LCD is operated in its 4 bit-mode. R/W (5) on the LCD MUST be hard wired
+ * low to prevent 5V logic appearing on the GPIO pins.
+ *
+ * Mappings can be set in the config file using the keys:
+ * pin_EN, pin_RS, pin_D7, pin_D6, pin_D5, pin_D4, pin_BL
+ * in the [hd44780] section.
+ */
+
+/*-
+ * Copyright (c) 2015 Fatih Aşıcı <fatih.asici@gmail.com>
+ *
+ * This file is released under the GNU General Public License. Refer to the
+ * COPYING file distributed with this package.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "hd44780-gpio.h"
+#include "hd44780-low.h"
+#include "report.h"
+
+void gpio_HD44780_senddata(PrivateData *p, unsigned char displayID,
+			   unsigned char flags, unsigned char ch);
+void gpio_HD44780_backlight(PrivateData *p, unsigned char state);
+void gpio_HD44780_close(PrivateData *p);
+
+typedef struct {
+	const char *name;
+	int number;
+	int fd;
+	int last_value;
+} gpio_pin;
+
+typedef struct {
+	gpio_pin en;
+	gpio_pin rs;
+	gpio_pin d7;
+	gpio_pin d6;
+	gpio_pin d5;
+	gpio_pin d4;
+	gpio_pin bl;
+} gpio_pins;
+
+
+/**
+ * Ask Linux kernel to export node for a GPIO.
+ *
+ * \param num   Number of the GPIO pin.
+ * \return      0 on success; -1 on error.
+ */
+static int
+export_gpio(int num)
+{
+	FILE *f = fopen("/sys/class/gpio/export", "w");
+	int ret;
+
+	if (f == NULL)
+		return -1;
+
+	fprintf(f, "%d", num);
+	ret = fclose(f);
+
+	/* If EBUSY is returned, this GPIO might be exported already.
+	 * Return failure for other errors. */
+	if (ret == EOF && errno != EBUSY)
+		return -1;
+
+	return 0;
+}
+
+
+/**
+ * Ask Linux kernel to remove a GPIO node exported previously.
+ *
+ * \param num   Number of the GPIO pin.
+ * \return      0 on success; -1 on error.
+ */
+static int
+unexport_gpio(int num)
+{
+	FILE *f = fopen("/sys/class/gpio/unexport", "w");
+
+	if (f == NULL)
```

Comment:
Add a perror() here. In my testing I get *Permission denied* while lcdproc tries to unexport every GPIO pin during shutdown.